### PR TITLE
Initial implementation of new self_test

### LINKED ISSUE #3701: Rethinking and rewriting sunpy.self_test
We are currently using astropy's test runner for `sunpy.self_test` this was really designed for setup.py and is therefore very full of features which are probably not needed for self_test.

Before we (I) go deleting swathes of code as I love to do. What do we want to achieve with self test? Is a very slim wrapper around `pytest --pyargs sunpy` all we need?

### LINKED ISSUE #4612: Can no longer run test suite without all the plugins because of warnings
This includes at least, doctestplus, asdf and remote-data

### PR DESCRIPTION
**Pull request by Cadair**
This kills the old runner and also adds a dependancy check before the test suite is run.

Fixes #4612
Fixes #3701

### CODE REVIEW

File: tox.ini [hunk 1]
```diff
@@ -108,6 +109,20 @@ commands =
     pre-commit install-hooks
     pre-commit run --verbose --all-files --show-diff-on-failure
 
+[testenv:base_deps]
```
Comment (nabobalis) [inline tox.ini:112]: Do we want to add this to the CI?

File: sunpy/tests/tests/test_self_test.py [hunk 2]
```diff
@@ -100,38 +83,6 @@ def test_main_figure_dir(monkeypatch):
                     [root_dir, '--figure_dir', '.', '-m', 'figure'])
 
 
-def test_main_coverage(monkeypatch):
```
Comment (nabobalis) [inline sunpy/tests/tests/test_self_test.py:103]: I removed coverage since we don't actually support it normally, only via the CI.

File: sunpy/tests/tests/test_self_test.py [hunk 3]
```diff
@@ -79,13 +69,6 @@ def test_main_only_remote_data(monkeypatch):
                     [os.path.join(root_dir, 'map'), '-k remote_data', '--remote-data=any', '-m', 'not figure'])
 
 
-def test_main_figures(monkeypatch):
```
Comment (nabobalis) [inline sunpy/tests/tests/test_self_test.py:82]: I removed `figures` and left it as `figure_only`

File: sunpy/tests/self_test.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+            importlib.import_module(f"sunpy.{package}")
+        except ModuleNotFoundError:
+            raise ModuleNotFoundError(f"sunpy.{package} was not found.")
+        args.extend(["--pyargs", f"sunpy.{package}"])
+    else:
+        args.extend(["--pyargs", "sunpy"])
+    if figure_only:
+        args.extend(["-m", "figure"])
+    if figure_dir:
+        args.extend(["--figure_dir", f"{figure_dir}"])
```
Comment (Cadair) [inline sunpy/tests/self_test.py:110]: pytest-mpl removed this, we should purge it.

File: sunpy/tests/self_test.py [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    if package:
+        try:
+            importlib.import_module(f"sunpy.{package}")
+        except ModuleNotFoundError:
+            raise ModuleNotFoundError(f"sunpy.{package} was not found.")
+        args.extend(["--pyargs", f"sunpy.{package}"])
+    else:
+        args.extend(["--pyargs", "sunpy"])
+    if figure_only:
+        args.extend(["-m", "figure"])
```
Comment (Cadair) [inline sunpy/tests/self_test.py:108]:
```suggestion
        args.extend(["-m", "mpl_image_compare"])
```

File: tox.ini (hunk 1 above)
Comment (Cadair) [reply tox.ini:112]: no? Could add it as cron only?

File: tox.ini [hunk 6]
```diff
@@ -108,6 +109,20 @@ commands =
...
+extras =
+deps =
+  numpy>=1.16.0
+  scipy>=1.2.0
+  matplotlib>=2.2.2
+  pandas>=0.23.0
+  astropy>=4.0
+  parfive[ftp]>=1.1.0
+commands =
+    python -c "import sunpy; sunpy.self_test()"
```
Comment (nabobalis) [inline tox.ini:124]: This now prints a nice message.

File: sunpy/tests/self_test.py (hunk 5 above)
Comment (nabobalis) [reply sunpy/tests/self_test.py:108]: Ah I forgot!

File: sunpy/tests/self_test.py [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+        printed = True
+        print(
+            f"The following packages are not installed for the {package}[{extra_name}] requirement:")
+        for dep in dependancies:
+            print(f"  * {dep}")
+
+    return printed
+
+
+def _self_test_args(*, package=None, online=False, online_only=False, figure_only=False):
```
Comment (nabobalis) [inline sunpy/tests/self_test.py:72]: This was my workaround to my failed attempt to get the monkey patching working.

**Comment by nabobalis** [comment]
I broke the conda build with these changes.

```python
===================================================================================== FAILURES =====================================================================================
___________________________________________________________________________ test_main_nonexisting_module ___________________________________________________________________________
[gw3] linux -- Python 3.8.6 /home/vsts/work/1/s/.tox/py38-conda/bin/python

    def test_main_nonexisting_module():
        with pytest.raises(ModuleNotFoundError):
>           sunpy.self_test(package='doesnotexist')
E           Failed: DID NOT RAISE <class 'ModuleNotFoundError'>

../../.tox/py38-conda/lib/python3.8/site-packages/sunpy/tests/tests/test_self_test.py:12: Failed
_____________________________________________________________________________ test_main_stdlib_module ______________________________________________________________________________
[gw3] linux -- Python 3.8.6 /home/vsts/work/1/s/.tox/py38-conda/bin/python

    def test_main_stdlib_module():
        """
        This test makes sure that the module is really searched within the sunpy
        package.
        """
        with pytest.raises(ModuleNotFoundError):
>           sunpy.self_test(package='random')
E           Failed: DID NOT RAISE <class 'ModuleNotFoundError'>
```

**Comment by nabobalis** [comment]
base_deps prints out:
```bash
Starting sunpy self test...
Checking for installed packages:
The following packages are not installed for the sunpy[asdf] requirement:
  * asdf
The following packages are not installed for the sunpy[dask] requirement:
  * dask
The following packages are not installed for the sunpy[database] requirement:
  * sqlalchemy
The following packages are not installed for the sunpy[image] requirement:
  * scikit-image
The following packages are not installed for the sunpy[jpeg2000] requirement:
  * glymur
The following packages are not installed for the sunpy[net] requirement:
  * beautifulsoup4
  * zeep
  * drms
You do not have all the required dependencies installed to run the sunpy test suite.
If you want to run the sunpy tests install the 'tests' extra with `pip install sunpy[all,tests]`
_________________________________________________________________________________________________________________________________________________________ summary __________________________________________________________________________________________________________________________________________________________
  base_deps: commands succeeded
  congratulations :)
```

**Comment by nabobalis** [comment]
> I broke the conda build with these changes.
> 
> ```python
> ===================================================================================== FAILURES =====================================================================================
> ___________________________________________________________________________ test_main_nonexisting_module ___________________________________________________________________________
> [gw3] linux -- Python 3.8.6 /home/vsts/work/1/s/.tox/py38-conda/bin/python
> 
>     def test_main_nonexisting_module():
>         with pytest.raises(ModuleNotFoundError):
> >           sunpy.self_test(package='doesnotexist')
> E           Failed: DID NOT RAISE <class 'ModuleNotFoundError'>
> 
> ../../.tox/py38-conda/lib/python3.8/site-packages/sunpy/tests/tests/test_self_test.py:12: Failed
> _____________________________________________________________________________ test_main_stdlib_module ______________________________________________________________________________
> [gw3] linux -- Python 3.8.6 /home/vsts/work/1/s/.tox/py38-conda/bin/python
> 
>     def test_main_stdlib_module():
>         """
>         This test makes sure that the module is really searched within the sunpy
>         package.
>         """
>         with pytest.raises(ModuleNotFoundError):
> >           sunpy.self_test(package='random')
> E           Failed: DID NOT RAISE <class 'ModuleNotFoundError'>
> ```

I can't replicate this locally with my conda test.

File: setup.cfg [hunk 8]
```diff
@@ -83,18 +83,17 @@ testpaths = "sunpy" "docs"
 norecursedirs = ".tox" "build" "docs[\/]_build" "docs[\/]generated" "*.egg-info" "examples" "sunpy[/\]_dev" ".jupyter" ".history" "tools" "sunpy[\/]extern"
 doctest_plus = enabled
 doctest_optionflags = NORMALIZE_WHITESPACE FLOAT_CMP ELLIPSIS
-addopts = --doctest-rst -m "not figure" --doctest-ignore-import-errors
+addopts = --doctest-rst -m "not mpl_image_compare" --doctest-ignore-import-errors
```
Comment (Cadair) [inline setup.cfg:86]: I don't think we need this. pytest-mpl skips if the `--mpl` flag isn't specified. As long as we can assume pytest-mpl is installed we don't need this. If we don't want to assume that then it's probably better to keep it.

File: sunpy/tests/self_test.py [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+        args.extend(["-m", "mpl_image_compare"])
+    return args
+
+
+def self_test(*, package=None, online=False, online_only=False, figure_only=False):
+    print("\n\n")
+    print("Starting sunpy self test...")
+    print("Checking for installed packages:")
+
+    missing = missing_dependencies_by_extra(exclude_extras=("dev", "all", "docs"))
```
Comment (Cadair) [inline sunpy/tests/self_test.py:96]: Should we exclude `dask` and maybe `asdf` from here? They aren't actually needed for any of the sunpy API, and I worry the print out might make people think they need them?

File: sunpy/tests/self_test.py [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+        args.extend(["--pyargs", "sunpy"])
+    if figure_only:
+        args.extend(["-m", "mpl_image_compare"])
+    return args
+
+
+def self_test(*, package=None, online=False, online_only=False, figure_only=False):
+    print("\n\n")
+    print("Starting sunpy self test...")
+    print("Checking for installed packages:")
```
Comment (Cadair) [inline sunpy/tests/self_test.py:94]: If we did exclude dask and asdf, we should probably reword this to say something like "checking for packages needed to run sunpy". I think we should draw a firmer line between the runtime and test deps in the output of this function.

File: sunpy/tests/self_test.py [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+
+    missing = missing_dependencies_by_extra(exclude_extras=("dev", "all", "docs"))
+    test_missing = missing.pop("tests")
+    printed = print_missing_dependencies_report(missing)
+
+    if not printed:
+        print("All required and optional sunpy dependencies are installed.")
+
+    if test_missing:
+        print("You do not have all the required dependencies installed to run the sunpy test suite.")
```
Comment (Cadair) [inline sunpy/tests/self_test.py:104]: I think this also probably needs rewording with a caveat like "you don't have the test deps but this doesn't mean sunpy wont work".

File: tox.ini [hunk 12]
```diff
@@ -107,6 +108,20 @@ commands =
...
```
Comment (Cadair) [inline tox.ini:123]: Are we actually verifying anything with this test though? What exit status do we return if deps are missing?

**Review by Cadair — COMMENTED** [review]
Looks good. Thanks for taking it over @nabobalis.

I worry that the copy from the self test function could be misleading in a bunch of places. I think the main users of this function are going to be really new people who want to make sure sunpy is installed correctly and people who are running into issues.

We should make sure the ordering of the output is a not technical as we can get away with and also that it draws a very firm line between "we recommend you install this" and "shit is broken because you don't have this installed".

I also think that everyone running this function should be assumed to be a regular user of sunpy and therefore be expected to have all the deps installed, so if anything is missing (maybe excluding dask and asdf) then we should be telling them what they want to do to install it.

File: setup.cfg (hunk 8 above)
Comment (nabobalis) [reply setup.cfg:86]: Fixed.

File: sunpy/tests/self_test.py (hunk 11 above)
Comment (nabobalis) [reply sunpy/tests/self_test.py:104]: Fixed.

File: sunpy/tests/self_test.py (hunk 9 above)
Comment (nabobalis) [reply sunpy/tests/self_test.py:96]: Fixed.

File: sunpy/tests/self_test.py (hunk 10 above)
Comment (nabobalis) [reply sunpy/tests/self_test.py:94]: Fixed.

File: tox.ini (hunk 12 above)
Comment (nabobalis) [reply tox.ini:123]: Just wanted to make sure it worked and using tox seemed the easiest thing to do.

Comment (nabobalis) [reply tox.ini:123]: LIke so;
```
.package create: /home/nabil/GitHub/sunpy/.tox/.package
.package installdeps: setuptools, setuptools_scm, wheel, oldest-supported-numpy, extension-helpers
base_deps create: /home/nabil/GitHub/sunpy/.tox/base_deps
base_deps installdeps: numpy>=1.16.0, scipy>=1.2.0, matplotlib>=2.2.2, pandas>=0.23.0, astropy>=4.0, parfive[ftp]>=1.1.0
base_deps inst: /home/nabil/GitHub/sunpy/.tox/.tmp/package/1/sunpy-2.1.dev782+g9b4b4adc7.d20201121.tar.gz
base_deps installed: aioftp==0.18.1,aiohttp==3.7.3,astropy==4.1,async-timeout==3.0.1,attrs==20.3.0,chardet==3.0.4,cycler==0.10.0,h5netcdf==0.8.1,h5py==3.1.0,idna==2.10,kiwisolver==1.3.1,matplotlib==3.3.3,multidict==5.0.2,numpy==1.19.4,pandas==1.1.4,parfive==1.2.0,Pillow==8.0.1,pyparsing==2.4.7,python-dateutil==2.8.1,pytz==2020.4,scipy==1.5.4,six==1.15.0,sunpy @ file:///home/nabil/GitHub/sunpy/.tox/.tmp/package/1/sunpy-2.1.dev782%2Bg9b4b4adc7.d20201121.tar.gz,tqdm==4.52.0,typing-extensions==3.7.4.3,yarl==1.6.3
base_deps run-test-pre: PYTHONHASHSEED='1235638279'
base_deps run-test: commands[0] | python -c 'import sunpy; sunpy.self_test()'



Starting sunpy self test...
Checking for packages needed to run sunpy:
The following packages are not installed for the sunpy[database] requirement:
  * sqlalchemy
The following packages are not installed for the sunpy[image] requirement:
  * scikit-image
The following packages are not installed for the sunpy[jpeg2000] requirement:
  * glymur
The following packages are not installed for the sunpy[net] requirement:
  * drms
  * beautifulsoup4
  * zeep
You do not have all the required dependencies installed to run the sunpy test suite.
NOTE THIS JUST MEANS THE TESTS WILL NOT RUN, NOT THAT SUNPY MISCONFIGURED.
If you want to run the sunpy tests install the 'tests' extra with `pip install sunpy[all,tests]`
_________________________________________________________________________________________________________________________________________________________ summary __________________________________________________________________________________________________________________________________________________________
  base_deps: commands succeeded
  congratulations :)
```

File: sunpy/tests/self_test.py [hunk 13]
```diff
@@ -0,0 +1,116 @@
+import importlib
+
+from pkg_resources import get_distribution
+
+
+def find_missing_dependencies(package="sunpy", extras=None):
```
Comment (nabobalis) [inline sunpy/tests/self_test.py:6]: I feel like we should test these?

**Review by nabobalis — APPROVED** [review]
